File: src/web/magmac/app/compile/rule/FilterRule.ts
export class FilterRule {
	 FilterRule( filter : Filter,  childRule : Rule) : private {this.childRule=childRule;this.filter=filter;;}
	public static Symbol( childRule : Rule) : Rule {return new FilterRule( new SymbolFilter( ), childRule);;}
	public static Number( childRule : Rule) : Rule {return new FilterRule( new NumberFilter( ), childRule);;}
	public lex( input : String) : CompileResult<Node> {if(true){ return this.childRule.lex( input);;}if(true){ return CompileErrors.createStringError( this.filter.createMessage( ), input);;};}
	public generate( node : Node) : CompileResult<String> {return this.childRule.generate( node);;}
}

File: packages/compiler-java/src/main/java/magmac/app/compile/rule/FilterRule.java
package magmac.app.compile.rule;

import magmac.app.compile.error.CompileResult;
import magmac.app.compile.error.type.CompileErrors;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.filter.Filter;
import magmac.app.compile.rule.filter.NumberFilter;
import magmac.app.compile.rule.filter.SymbolFilter;

public final class FilterRule implements Rule {
    private final Rule childRule;
    private final Filter filter;

    private FilterRule(Filter filter, Rule childRule) {
        this.childRule = childRule;
        this.filter = filter;
    }

    public static Rule Symbol(Rule childRule) {
        return new FilterRule(new SymbolFilter(), childRule);
    }

    public static Rule Number(Rule childRule) {
        return new FilterRule(new NumberFilter(), childRule);
    }

    @Override
    public CompileResult<Node> lex(String input) {
        if (this.filter.test(input)) {
            return this.childRule.lex(input);
        }
        else {
            return CompileErrors.createStringError(this.filter.createMessage(), input);
        }
    }

    @Override
    public CompileResult<String> generate(Node node) {
        return this.childRule.generate(node);
    }
}
